django/django-react-integration.md

```diff
@@ -0,0 +1,281 @@
+
+### Define a view with a path to your component
+
+Define a Django view that includes a `component` attribute with a staticfile
+path to the component you've defined, as well as the typical `template_name`
+attribute representing the path to a Django template. The combination of these
+two attributes will allow us to render the component in the view's template.
+
+```python
+class Home(TemplateView):
+    title = 'Home'
+    template_name = 'example_app/index.html'
+    component = 'js/pages/index.js'  # Staticfile path to your component code
+```
+
+### Add a `props` key to your view context dictionary
+
+In the context dictionary for your view, add a `props` key representing the props
+that you want to pass into your React component. If you're using class-based views,
+you can update your context dictionary using the `get_context_data()` method.
+
+```python
+class Home(TemplateView):
+    title = 'Home'
+    template_name = 'example_app/index.html'
+    component = 'js/pages/index.js'  # Staticfile path to your component code
+
+    def get_context_data(self, *args, **kwargs):
+        context = super().get_context_data(*args, **kwargs)
+        context['props'] = {'user': 'foobar'}  # Set React props here
+        return context
+```
+
+### Create a Django template for your React component
+
+In order to render the React component we just defined, we need a Django template
+that can mark the component for bundling. Save this template to the path
+defined in your `template_name` view attribute.
+
+There are a few key components to this template. We'll consider them one by one.
+
+#### Define a React root element
+
+The template needs to have an element that React can use as a root to mount
+your component. For a better user experience, make sure to embed a loading animation
+like a spinner in this element so the user can see that your component is rendering.
+
+As an example, a simple root element might look like this:
+
+```html
+<div id="App">
+  <!-- Contents get replaced by mounted React.Component -->
+  <div class="text-center">
+    <i class="fa fa-lg fa-spinner fa-spin"></i><br><br>
+    <i class="pending">Loading components...</i><br><br>
+  </div>
+</div>
+```
+
+#### Set React props and root element on the `window` object
+
+In order to make your props and root element available to your React component, add
+a simple script to the end of your template to set them on the global `window`
+object.
+
+```html
+{{ props|json_script:"props" }}  // Serialize and safely sanitize the props
+<script type="text/javascript">
+  window.props = JSON.parse(document.getElementById('props').textContent)
+  window.reactMount = document.getElementById('App')
+</script>
+```
+
+#### Bundle and load your component with `django-compressor`
+
+Finally, use the `compress` templatetag provided by `django-compressor` and the
+`view.component` attribute to bundle and load your component as an external script.
```

Comment:
Where does `view.component` come from?

Reply:
It's set on the view for this template in the section `Define a view with a path to your component` above:

> Define a Django view that includes a `component` attribute with a staticfile
> path to the component you've defined, as well as the typical `template_name`
> attribute representing the path to a Django template. The combination of these
> two attributes will allow us to render the component in the view's template.
>
> ```python
> class Home(TemplateView):
>     title = 'Home'
>     template_name = 'my_new_app/index.html'
>     component = 'js/pages/index.js'  # Staticfile path to your component code
> ```

I'll adjust the wording here to make it a little bit clearer where `view.component` is coming from.